feat: add `blas/ext/base/gfind-index`

headlessNode opened the pull request:
---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes. report:
  - task: lint_filenames status: passed
  - task: lint_editorconfig status: passed
  - task: lint_markdown status: passed
  - task: lint_package_json status: passed
  - task: lint_repl_help status: passed
  - task: lint_javascript_src status: passed
  - task: lint_javascript_cli status: na
  - task: lint_javascript_examples status: passed
  - task: lint_javascript_tests status: passed
  - task: lint_javascript_benchmarks status: passed
  - task: lint_python status: na
  - task: lint_r status: na
  - task: lint_c_src status: na
  - task: lint_c_examples status: na
  - task: lint_c_benchmarks status: na
  - task: lint_c_tests_fixtures status: na
  - task: lint_shell status: na
  - task: lint_typescript_declarations status: passed
  - task: lint_typescript_tests status: passed
  - task: lint_license_headers status: passed ---

Progresses #2656.

## Description

> What is the purpose of this pull request?

This pull request:

-   add `blas/ext/base/gfind-index`

## Related Issues

> Does this pull request have any related issues?

This pull request:

-   progresses #2656

## Questions

> Any questions for reviewers of this pull request?

No.

## Other

> Any other information relevant to this pull request? This may include screenshots, references, and/or implementation notes.

No.

## Checklist

> Please ensure the following tasks are completed before submitting this pull request.

-   [x] Read, understood, and followed the [contributing guidelines][contributing].

* * *

@stdlib-js/reviewers

[contributing]: https://github.com/stdlib-js/stdlib/blob/develop/CONTRIBUTING.md

headlessNode committed:
feat: add blas/ext/base/gfind-index

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: passed
  - task: lint_package_json
    status: passed
  - task: lint_repl_help
    status: passed
  - task: lint_javascript_src
    status: passed
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: passed
  - task: lint_javascript_tests
    status: passed
  - task: lint_javascript_benchmarks
    status: passed
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: passed
  - task: lint_typescript_tests
    status: passed
  - task: lint_license_headers
    status: passed
---

headlessNode commented on lib/node_modules/@stdlib/blas/ext/base/gfind-index/README.md:
```diff
@@ -0,0 +1,229 @@
+
+#### gfindIndex( N, x, strideX, clbk\[, thisArg] )
+
+Returns the index of the first element which passes a test implemented by a predicate function.
```
```suggestion
Returns the first index of an element which passes a test implemented by a predicate function.
```
@kgryte Is this suggestion okay or the original description?

kgryte replied:
The original description is fine.

headlessNode commented on lib/node_modules/@stdlib/blas/ext/base/gfind-index/README.md:
```diff
@@ -0,0 +1,229 @@
+-   **N**: number of indexed elements.
+-   **x**: input array.
+-   **strideX**: stride length.
+-   **predicate**: predicate function.
```
```suggestion
-   **clbk**: callback function.
```

headlessNode commented on lib/node_modules/@stdlib/blas/ext/base/gfind-index/docs/repl.txt:
```diff
@@ -0,0 +1,105 @@
+        Callback function.
+
+    thisArg: any (optional)
+        Execution context.
```
```suggestion
        Callback execution context.
```

headlessNode commented on lib/node_modules/@stdlib/blas/ext/base/gfind-index/docs/repl.txt:
```diff
@@ -0,0 +1,105 @@
+        Callback function.
+
+    thisArg: any (optional)
+        Execution context.
```
```suggestion
        Callback execution context.
```

headlessNode committed:
docs: apply suggestions from code review

Signed-off-by: Muhammad Haris <101793258+headlessNode@users.noreply.github.com>

headlessNode commented on lib/node_modules/@stdlib/blas/ext/base/gfind-index/lib/accessors.js:
```diff
@@ -0,0 +1,75 @@
+* @param {NonNegativeInteger} offsetX - starting index
+* @param {Callback} clbk - callback function
+* @param {*} thisArg - execution context
+* @returns {Object} input array object
```
```suggestion
* @returns {integer} index
```

headlessNode commented on lib/node_modules/@stdlib/blas/ext/base/gfind-index/lib/main.js:
```diff
@@ -0,0 +1,56 @@
+* @param {integer} strideX - stride length
+* @param {Callback} clbk - callback function
+* @param {*} [thisArg] - execution context
+* @returns {Collection} input array
```
```suggestion
* @returns {integer} index
```

headlessNode commented on lib/node_modules/@stdlib/blas/ext/base/gfind-index/lib/ndarray.js:
```diff
@@ -0,0 +1,75 @@
+* @param {NonNegativeInteger} offsetX - starting index
+* @param {Callback} clbk - callback function
+* @param {*} [thisArg] - execution context
+* @returns {Collection} input array
```
```suggestion
* @returns {integer} index
```

headlessNode committed:
docs: apply suggestions from code review

Signed-off-by: Muhammad Haris <101793258+headlessNode@users.noreply.github.com>

kgryte commented on lib/node_modules/@stdlib/blas/ext/base/gfind-index/README.md:
````diff
@@ -0,0 +1,229 @@
+// returns 3
+```
+
+If the function is unable to find an element, the function returns `-1`.
````
```suggestion
If no element passes a test implemented by a predicate function, the function returns `-1`.
```

kgryte commented on lib/node_modules/@stdlib/blas/ext/base/gfind-index/README.md:
````diff
@@ -0,0 +1,229 @@
+// returns 4
+```
+
+The `N` and stride parameters determine which elements in the strided array are accessed at runtime. For example, to check every other element:
````
```suggestion
The `N` and stride parameters determine which elements in the strided array are accessed at runtime. For example, to test every other element:
```

kgryte commented on lib/node_modules/@stdlib/blas/ext/base/gfind-index/README.md:
```diff
@@ -0,0 +1,229 @@
+// Initial array...
+var x0 = new Float64Array( [ 1.0, 3.0, 4.0, -5.0, 7.0, 6.0 ] );
+
+// Create an offset view...
```
```suggestion
// Initial array:
var x0 = new Float64Array( [ 1.0, 3.0, 4.0, -5.0, 7.0, 6.0 ] );

// Create an offset view:
```

kgryte committed:
Apply suggestions from code review

Signed-off-by: Athan <kgryte@gmail.com>

kgryte commented on lib/node_modules/@stdlib/blas/ext/base/gfind-index/benchmark/benchmark.js:
```diff
@@ -0,0 +1,96 @@
+* @param {*} v - array element
+* @returns {boolean} result
+*/
+function isEven( v ) {
```
@headlessNode Always attempt to measure the worst case scenario. In this case, it's likely that you're testing a best case scenario (i.e., an even number is the first or second element). Rename `isEven` to `clbk`. Instead of `v % 2 === 0`, you can use `( v > 999.0 )` which is clearly beyond the range of the generated values.

kgryte commented on lib/node_modules/@stdlib/blas/ext/base/gfind-index/benchmark/benchmark.ndarray.js:
```diff
@@ -0,0 +1,96 @@
+* @param {*} v - array element
+* @returns {boolean} result
+*/
+function isEven( v ) {
```
Same comment.

kgryte commented on lib/node_modules/@stdlib/blas/ext/base/gfind-index/docs/repl.txt:
```diff
@@ -0,0 +1,105 @@
+
+    Indexing is relative to the first index. To introduce an offset, use typed
+    array views.
+
```
@headlessNode You should add a note in here regarding what happens when an element is not found.

kgryte commented on lib/node_modules/@stdlib/blas/ext/base/gfind-index/docs/types/index.d.ts:
```diff
@@ -0,0 +1,199 @@
+	* @returns `index`
+	*
+	* @example
+	*
```
```suggestion
```

kgryte committed:
style: remove empty line

Signed-off-by: Athan <kgryte@gmail.com>

kgryte commented on lib/node_modules/@stdlib/blas/ext/base/gfind-index/docs/types/index.d.ts:
```diff
@@ -0,0 +1,198 @@
+	* @param strideX - stride length
+	* @param clbk - callback function
+	* @param thisArg - execution context
+	* @returns `index`
```
`index` here and below does not need to be in backticks.

kgryte commented on lib/node_modules/@stdlib/blas/ext/base/gfind-index/docs/types/test.ts:
```diff
@@ -0,0 +1,191 @@
+
+// TESTS //
+
+// The function returns an integer...
```
```suggestion
// The function returns a number...
```

kgryte commented on lib/node_modules/@stdlib/blas/ext/base/gfind-index/lib/accessors.js:
```diff
@@ -0,0 +1,75 @@
+	// Cache reference to array data:
+	xbuf = x.data;
+
+	// Cache a reference to the element accessors:
```
```suggestion
	// Cache a reference to the element accessor:
```

kgryte commented on lib/node_modules/@stdlib/blas/ext/base/gfind-index/test/test.main.js:
```diff
@@ -0,0 +1,145 @@
+	}
+});
+
+tape( 'the function returns `-1` if provided `N` parameter is less than or equal to zero', function test( t ) {
+	var actual;
+
+	actual = gfindIndex( 0, [ 1.0, 2.0, 3.0 ], 1, clbk );
+	t.strictEqual( actual, -1, 'returns expected value' );
+
+	actual = gfindIndex( -1, [ 1.0, 2.0, 3.0 ], 1, clbk );
+	t.strictEqual( actual, -1, 'returns expected value' );
+
+	t.end();
+
+	function clbk( v ) {
+		return v % 2.0 === 0.0;
+	}
+});
+
+tape( 'the function returns `-1` if provided `N` parameter is less than or equal to zero (accessors)', function test( t ) {
```
```suggestion
tape( 'the function returns `-1` if the provided `N` parameter is less than or equal to zero', function test( t ) {
	var actual;

	actual = gfindIndex( 0, [ 1.0, 2.0, 3.0 ], 1, clbk );
	t.strictEqual( actual, -1, 'returns expected value' );

	actual = gfindIndex( -1, [ 1.0, 2.0, 3.0 ], 1, clbk );
	t.strictEqual( actual, -1, 'returns expected value' );

	t.end();

	function clbk( v ) {
		return v % 2.0 === 0.0;
	}
});

tape( 'the function returns `-1` if the provided `N` parameter is less than or equal to zero (accessors)', function test( t ) {
```

kgryte committed:
test: update descriptions

Signed-off-by: Athan <kgryte@gmail.com>

kgryte committed:
docs: update comment

Signed-off-by: Athan <kgryte@gmail.com>

kgryte committed:
docs: update comment

Signed-off-by: Athan <kgryte@gmail.com>

kgryte commented on lib/node_modules/@stdlib/blas/ext/base/gfind-index/test/test.ndarray.js:
```diff
@@ -0,0 +1,145 @@
+	x = [ 1.0, 1.0, 2.0, 3.0, 2.0, 3.0 ];
+
+	// Nonnegative stride...
+	actual = gfindIndex( x.length, x, 1, 3, clbk );
```
Your `N` parameters here and below are incorrect. It should not be `x.length` when the `offset` is non-zero.

kgryte commented on lib/node_modules/@stdlib/blas/ext/base/gfind-index/test/test.ndarray.js:
```diff
@@ -0,0 +1,145 @@
+	}
+});
+
+tape( 'the function returns `-1` if provided `N` parameter is less than or equal to zero (accessors)', function test( t ) {
```
```suggestion
tape( 'the function returns `-1` if the provided `N` parameter is less than or equal to zero', function test( t ) {
	var actual;

	actual = gfindIndex( 0, [ 1.0, 2.0, 3.0 ], 1, 0, clbk );
	t.strictEqual( actual, -1, 'returns expected value' );

	actual = gfindIndex( -1, [ 1.0, 2.0, 3.0 ], 1, 2, clbk );
	t.strictEqual( actual, -1, 'returns expected value' );

	t.end();

	function clbk( v ) {
		return v % 2.0 === 0.0;
	}
});

tape( 'the function returns `-1` if the provided `N` parameter is less than or equal to zero (accessors)', function test( t ) {
```

kgryte committed:
test: update descriptions

Signed-off-by: Athan <kgryte@gmail.com>

kgryte reviewed (changes requested):
After some clean-up, mainly minor things remain. Namely the benchmarks and cleaning up the tests.

headlessNode committed:
fix: benchmarks

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: na
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: passed
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

headlessNode committed:
docs: fix repl docs

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: passed
  - task: lint_javascript_src
    status: na
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: passed
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

headlessNode committed:
test: fix incorrect offsets

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: na
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: passed
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

kgryte commented on lib/node_modules/@stdlib/blas/ext/base/gfind-index/docs/repl.txt:
```diff
@@ -0,0 +1,106 @@
+
+    Examples
+    --------
+    // Standard Usage:
```
```suggestion
```

kgryte committed:
docs: remove comment

Signed-off-by: Athan <kgryte@gmail.com>

kgryte reviewed (approved):
LGTM
